src/vs/base/browser/ui/tree/stickyScroll.ts

```diff
@@ -0,0 +1,459 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import 'vs/css!./media/stickyScroll';
+import { AbstractTree, IAbstractTreeOptions, IAbstractTreeOptionsUpdate, TreeRenderer } from 'vs/base/browser/ui/tree/abstractTree';
+import { ITreeModel, ITreeNode } from 'vs/base/browser/ui/tree/tree';
+import { Disposable, IDisposable, toDisposable } from 'vs/base/common/lifecycle';
+import { List, isActionItem, isMonacoCustomToggle } from 'vs/base/browser/ui/list/listWidget';
+import { equals } from 'vs/base/common/arrays';
+import { IListVirtualDelegate } from 'vs/base/browser/ui/list/list';
+import { $, addDisposableListener } from 'vs/base/browser/dom';
+
+interface StickyScrollNode<T, TFilterData> {
+	readonly node: ITreeNode<T, TFilterData>;
+	readonly startIndex: number;
+	readonly endIndex: number;
+	readonly height: number;
+	readonly position: number;
+}
+
+class StickyScrollState<T, TFilterData, TRef> extends Disposable {
```

Comment:
Let's just put this next to the filter stuff in `abstractTree.ts`. That file is getting huge, sure... but if we want to clean it, let's do it properly in a next step, instead of creating a semi-done state.